src/pool/srv_pool_scrub.c

```diff
@@ -0,0 +1,345 @@
+ * but for now, to make testing easier using environment variables to configure.
+ */
+static bool
+scrubbing_is_enabled()
+{
+	char *enabled = getenv("DAOS_CSUM_SCRUB");
+
+	return enabled != NULL && strncmp(enabled, "ON", strlen("ON")) == 0;
+}
+
+static uint64_t
+interval_sec()
+{
+	char *sec = getenv("DAOS_CSUM_SCRUB_INTERVAL_SEC");
+
+	return sec != NULL ? atoll(sec) : 7 * 24 * 60 * 60;
+}
+
+static uint64_t
+between_scrub_sec()
+{
+	char *sec = getenv("DAOS_CSUM_SCRUB_PAUSE_SEC");
+
+	return sec != NULL ? atoll(sec) : 24 * 60 * 60;
+}
+
+/** vos_iter_cb_t */
+static int
+obj_iter_scrub_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+		  vos_iter_type_t type, vos_iter_param_t *param,
+		  void *cb_arg, unsigned int *acts)
+{
+	struct scrub_ctx	*ctx = cb_arg;
+	struct daos_csummer	*csummer = ctx->cur_cont->sc_csummer;
+
+	if (!(type == VOS_ITER_RECX || type == VOS_ITER_SINGLE))
+		return 0;
+
+	C_TRACE("Scrubbing akey: "DF_KEY", type: %s, record size: "
+			DF_U64", extent: "DF_RECX", csum_size: %d\n",
+		DP_KEY(&param->ip_akey),
+		(type == VOS_ITER_RECX) ? "ARRAY" : "SV",
+		entry->ie_rsize,
+		DP_RECX(entry->ie_orig_recx),
+		daos_csummer_get_csum_len(csummer)
+	);
+
+	/** TODO - implement the actual scrubbing of the checksum by fetching
+	 * the data, calculating a new checksum, and comparing to the stored
+	 * checksum.
+	 */
+	ctx->pool_csums_scrubbed++;
+
+	if (ctx->msec_between_calcs == 0) {
+		C_TRACE("Yield after data scrub\n");
+		dss_ult_yield(ctx->req);
+	} else {
+		C_TRACE("Sleeping after data scrub for "DF_U64" msec\n",
+			ctx->msec_between_calcs);
+		sched_req_sleep(ctx->req, ctx->msec_between_calcs);
+	}
+
+	return 0;
+}
+
+static int
+scrub_cont(struct scrub_ctx *ctx)
+{
+	vos_iter_param_t		param = {0};
+	struct vos_iter_anchors		anchor = {0};
+	int				rc;
+
+	/** Do the actual scrubbing */
+	C_TRACE("Scrubbing container '"DF_UUIDF"'\n",
+		DP_UUID(ctx->cur_cont->sc_uuid));
+
+	param.ip_hdl = ctx->cur_cont->sc_hdl;
+	param.ip_epr.epr_lo = 0;
+	param.ip_epr.epr_hi = DAOS_EPOCH_MAX;
+
+	rc = vos_iterate(&param, VOS_ITER_OBJ, true, &anchor,
+			 obj_iter_scrub_cb,
+			 NULL, ctx, NULL);
+
+	if (rc != DER_SUCCESS) {
+		D_ERROR("Object scrub failed: "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	return 0;
+}
+
+/** vos_iter_cb_t */
+static int
+cont_iter_scrub_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+		   vos_iter_type_t type, vos_iter_param_t *param,
+		   void *cb_arg, unsigned int *acts)
+{
+	struct scrub_ctx	*ctx = cb_arg;
+	struct ds_cont_child	*cont = NULL;
+	int			 rc = 0;
+
+	D_ASSERT(type == VOS_ITER_COUUID);
+
+	/** get per-thread container object which has the csummer */
+	rc = ds_cont_child_lookup(ctx->pool_uuid, entry->ie_couuid, &cont);
+	if (rc != DER_SUCCESS) {
+		D_ERROR("Lookup container '"DF_UUIDF"' failed: "DF_RC"\n",
+			DP_UUID(entry->ie_couuid), DP_RC(rc));
+		return rc;
+	}
+
+	if (daos_csummer_initialized(cont->sc_csummer)) {
+		ctx->cur_cont = cont;
+		rc = scrub_cont(ctx);
```

Comment:
Please be aware that if ULT yield in scrub_cont(), you have to inform vos_iterate() to re-probe by VOS_ITER_CB_YIELD flag.

Actually you can do a single recursive vos_iterate() call from container level, instead of call a non-recursive vos_iterate() in container level, then call another recursive vos_iterate() in object level.